cbmc/proofs/mqtt_state.c

```diff
@@ -285,41 +285,64 @@ bool valid_IotMqttNetworkInfo( const IotMqttNetworkInfo_t *pInfo )
  * IotMqttConnectInfo
  ****************************************************************/
 
+void bound_IotMqttConnectInfo( IotMqttConnectInfo_t *pInfo )
+{
+#ifdef SUBSCRIPTION_COUNT_MAX
+  __CPROVER_assume( pInfo->previousSubscriptionCount < SUBSCRIPTION_COUNT_MAX );
+#endif
+
+#ifdef CLIENT_IDENTIFIER_LENGTH_MAX
+  __CPROVER_assume( pInfo->clientIdentifierLength < CLIENT_IDENTIFIER_LENGTH_MAX &&
+                    VALID_CBMC_SIZE( pInfo->clientIdentifierLength ) );
+#else
+  __CPROVER_assume( pInfo->clientIdentifierLength <= UINT16_MAX );
+#endif
+
+#ifdef USER_NAME_LENGTH_MAX
+    __CPROVER_assume( pInfo->userNameLength < USER_NAME_LENGTH_MAX &&
+                      VALID_CBMC_SIZE( pInfo->userNameLength ) );
+#else
+  __CPROVER_assume( pInfo->userNameLength <= UINT16_MAX );
+#endif
+
+#ifdef PASSWORD_LENGTH_MAX
+    __CPROVER_assume( pInfo->passwordLength < PASSWORD_LENGTH_MAX &&
+                      VALID_CBMC_SIZE( pInfo->passwordLength ) );
+#else
+  __CPROVER_assume( pInfo->passwordLength <= UINT16_MAX );
+#endif
+}
+
 IotMqttConnectInfo_t *allocate_IotMqttConnectInfo( IotMqttConnectInfo_t *pInfo )
 {
   if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
   if ( pInfo == NULL ) return NULL;
 
+  bound_IotMqttConnectInfo( pInfo );
```

Comment:
This is a list of assumptions.  It doesn't belong in the allocators.  It belongs in one of the isvalid predicates.

Reply:
Before, we still needed assumptions in the `allocate_*` function such as (see `allocate_IotMqttPublishInfo`)
```c
  // assumption here is checked below in valid_
  __CPROVER_assume( VALID_CBMC_SIZE( pInfo->payloadLength ) );
```

To avoid the problem of duplicating assumptions when trying to allocate memory using an unconstrained size which is later bounded in the code (i.e., in the validity function).

We base our style in the AWS C Common proofs. These proofs use 3 predicates: "bounded" where we bound all necessary elements of the data structure, "allocated" where we allocate memory based on the bounds previously defined, and "validity" where we assume the invariant for the data structure. I like this style because it allow us to separate the assumptions about the bound and the actual invariant/properties for the data structure as well as avoid problems of duplicating assumptions when trying to allocate memory of unconstrained size (like the one we have in `allocate_IotMqttPublishInfo`).

In AWS C Common, we call "bounded" before the "allocated" predicate, but this would require to pass the bounded data structure as a parameter (instead of NULL) and I'd need to update other MQTT proofs. We can avoid that for now by keeping the call inside `allocate_*` functions.

I find this style cleaner, it is closer to what we do in AWS C Common, and it didn't break any proofs (neither decrease coverage), thus, I'd keep it and leave the discussion to apply this across all MQTT state to our second round.

Reply:
OK, this is a straight refactoring.